### DISCUSSION TITLE
first selection

### DISCUSSION THREAD

**Discussion by semi710** [hrsh7th/nvim-cmp#1670] [Q&A]
for some reason if i type my cmp highlights the selection in the middle of the list.
![image](https://github.com/hrsh7th/nvim-cmp/assets/60490474/3184cebd-ff16-40d2-9928-ba840449f549)

what i want is that it should not select any of the entry by default at all and if i press control l it should just select the top completion or if i manully scroll any entry the pressing control l selects that

**Comment by andreykaipov** [comment]
Had the same issue. Setting `preselect = cmp.PreselectMode.None` got the intended behaviour.

**Reply by semi710** [reply]
can you please tell where to put this option exactly?

**Reply by andreykaipov** [reply]
Oh sorry it's one of the top-level keys on setup:

```lua
cmp.setup({
    preselect = cmp.PreselectMode.None
    mapping = cmp.mapping.preset.insert({
        ...
    })
})
```

**Reply by itaranto** [reply]
Do you know the reason why this isn't the default?

I've just wasted several hours trying to fix this until I found this discussion.

The docs say:

```preselect 
  `cmp.PreselectMode`

  1. `cmp.PreselectMode.Item`
    nvim-cmp will preselect the item that the source specified.
  2. `cmp.PreselectMode.None`
    nvim-cmp will not preselect any items.
```

I wonder what _"will preselect the item that the source specified."_ means. I don't see any further documentation about this and seems to be a pretty big deal.

**Reply by itaranto** [reply]
Oh, ignore my previous comment.

I don't actually think this actually fixes OP's question, or at least why I think he asked first.

He asked why the selection is on the _middle_ of the list. 

This is also happening to me, LSP is preselected always regardless of priority.

**Reply by semi710** [reply]
My selection is still in the middle of the list.
When I marked it as an answer, it happened to be a coincidence that the choices were at the top.

**Reply by itaranto** [reply]
Yes, but `cmp.PreselectMode.None` actually makes no completion is selected by default, so it's kind of a bad way of fixing the problem.

Ideally the first entry on the list should be pre-selected.

**Comment by fitrh** [comment]
There was an issue for `gopls` about preselect completion, so in addition to setting `preselect` to `cmp.PreselectMode.None`, you also need to set `completion.completeopt` to `noselect`

```lua
cmp.setup({
  preselect = cmp.PreselectMode.None,
  completion = {
    completeopt = 'menu,menuone,noselect',
  },
})
```

If you want to avoid LSP preselect but stil want the vim preselect feature, you can set your `completeopt` without `noselect`


```lua
cmp.setup({
  preselect = cmp.PreselectMode.None,
  completion = {
    completeopt = 'menu,menuone,preview',
  },
})
```

**Reply by itaranto** [reply]
OK, but how coould he LSP have influence over this? Sources are managed by `nvim-cmp` as far as I understand. 

I'm not talking about different completion types from the same source, I'm talking about completely different nvim-cmp soruces, like `nvim_lsp` vs `luasnip`.


About using `cmp.PreselectMode.None`, that's definitely not ideal since I now need to press an extra key to select the first entry.

**Reply by fitrh** [reply]
> OK, but how coould he LSP have influence over this? Sources are managed by nvim-cmp as far as I understand.
>
> I'm not talking about different completion types from the same source, I'm talking about completely different nvim-cmp soruces, like nvim_lsp vs luasnip.

Can you share your cmp config? And maybe a reproduction steps of your issue?

> About using cmp.PreselectMode.None, that's definitely not ideal since I now need to press an extra key to select the first entry.

You can use `cmp.confirm({ select = true })` for your confirm key to automatically select the first item without preselect

**Reply by itaranto** [reply]
> Can you share your cmp config? And maybe a reproduction steps of your issue?

Here's my config:

```lua
    cmp.setup({
      formatting = {
        format = function(entry, vim_item)
          vim_item.menu = ({
            nvim_lsp = '[LSP]',
            nvim_lua = '[Lua]',
            luasnip = '[Snippet]',
            calc = '[Calc]',
            path = '[Path]',
            spell = '[Spell]',
            buffer = '[Buffer]',
          })[entry.source.name]
          return vim_item
        end,
      },
      mapping = {
        ['<C-p>'] = cmp.mapping.select_prev_item(),
        ['<C-n>'] = cmp.mapping.select_next_item(),
        ['<C-d>'] = cmp.mapping.scroll_docs(4),
        ['<C-u>'] = cmp.mapping.scroll_docs(-4),
        ['<CR>'] = cmp.mapping.confirm({
          behavior = cmp.ConfirmBehavior.Insert,
          select = true,
        }),
        ['<Tab>'] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_next_item()
          elseif luasnip.expand_or_locally_jumpable() then
            luasnip.expand_or_jump()
          else
            fallback()
          end
        end, { 'i', 's' }),
        ['<S-Tab>'] = cmp.mapping(function(fallback)
          if cmp.visible() then
            cmp.select_prev_item()
          elseif luasnip.locally_jumpable(-1) then
            luasnip.jump(-1)
          else
            fallback()
          end
        end, { 'i', 's' }),
      },
      snippet = {
        expand = function(args) luasnip.lsp_expand(args.body) end,
      },
      sources = cmp.config.sources({
        { name = 'luasnip'  },
        { name = 'nvim_lsp' },
        { name = 'nvim_lua' },
        { name = 'calc'     },
        { name = 'path'     },
        {
          name = 'spell',
          option = {
            keep_all_entries = true,
          },
        },
      }, {
        {
          name = 'buffer',
          option = {
            get_bufnrs = function()
              local bufs = {}
              for _, win in ipairs(vim.api.nvim_list_wins()) do
                bufs[vim.api.nvim_win_get_buf(win)] = true
              end
              return vim.tbl_keys(bufs)
            end,
          },
        },
      }),
    })
```

My issue is simple:
- Open a `.go` file
- Type `if`
- The following completion menu appears
![image](https://github.com/hrsh7th/nvim-cmp/assets/158330647/dd6d70b8-fed5-4f2e-84c4-015934967ed1)

**Reply by fitrh** [reply]
Yes, this is still the same issue I mentioned before, it is a `gopls` specific problem, there is an issue in `gopls` repo but I can't remember it, you need to add the following config

```lua
preselect = cmp.PreselectMode.None,
completion = {
  completeopt = 'menu,menuone,preview',
}
```

The `preselect` is to avoid LSP preselect feature and the `complete.completeopt` to enable vim preselect feature, so now you ignore LPS preselect but always get presect item from vim

**Reply by fitrh** [reply]
I have used this config (without preselect) for years with `gopls` without an issue, and this is with vim preselect enabled

https://github.com/hrsh7th/nvim-cmp/assets/55179750/b37a89e2-749b-4f70-8152-fe54bb434cb2

**Reply by itaranto** [reply]
OK, that works.

I didn't know about the LSP preselect feature.

Also, it seems I need to set `completeopt` through `nvim-cmp` otherwise it doesn't work.

I mean, `vim.opt.completeopt = 'menu,menuone,preview'` doesn't work but `completion.completeopt = ...` does.

**Reply by fitrh** [reply]
> Also, it seems I need to set completeopt through nvim-cmp otherwise it doesn't work.

That's right, you need to use cmp `completion.completeopt`

**Reply by fitrh** [reply]
@niksingh710 let me know if you still have the issue after this, also, in my experience, this is only happens with `gopls`

**Reply by semi710** [reply]
@fitrh yep I have noticed that this only happens in go, not in other languages.
